Diff editor

MrFlashAccount opened the pull request:
### Pull Request Description

Closes  enso-org/cloud#940

<!--
- Please describe the nature of your PR here, as well as the motivation for it.
- If it fixes an open issue, please mention that issue number here.
-->

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

MrFlashAccount committed:
Implement diff editor

MrFlashAccount committed:
Rebase over the latest dev

MrFlashAccount committed:
Fix eslint

MrFlashAccount committed:
fix build

MrFlashAccount commented on app/ide-desktop/eslint.config.js:
```diff
@@ -228,11 +222,6 @@ const RESTRICTED_SYNTAXES = [
         selector: 'CallExpression[callee.name=toastAndLog][arguments.0.value=/\\.$/]',
         message: '`toastAndLog` already includes a trailing `.`',
     },
-    {
```
Replaced with `native` rule from the 'react-eslint`

MrFlashAccount commented on app/ide-desktop/eslint.config.js:
```diff
@@ -290,7 +279,7 @@ export default [
             ...tsEslint.configs['recommended-requiring-type-checking']?.rules,
             ...tsEslint.configs.strict?.rules,
-            ...react.configs.recommended.rules,
+            ...react.configs['jsx-runtime'].rules,
```
Starting from react 17 we don't need to import React in each component, so I replaced the `recommended` with `jsx-runtime`

somebody1234 commented on app/ide-desktop/lib/assets/close.svg:
```diff
@@ -3,4 +3,4 @@
     <circle cx="8" cy="8" r="7" stroke="black" stroke-width="2" />
-</svg>
\ No newline at end of file
+</svg>
```
unintentional change?

MrFlashAccount replied:
Why did you mention this?

somebody1234 replied:
it makes the PR touch unnecessary files

(edit: more importantly, is there a tangible benefit to removing the trailing newline?)

MrFlashAccount replied:
The IDE might did that for me.

somebody1234 commented on app/ide-desktop/lib/assets/close_large.svg:
```diff
@@ -1,7 +1,7 @@
     </g>
-</svg>
\ No newline at end of file
+</svg>
```
another unintentional change?

somebody1234 commented on app/ide-desktop/lib/dashboard/src/App.tsx:
```diff
@@ -185,6 +190,9 @@ function AppRouter(props: AppProps) {
     window.navigate = navigate
   }
   const [inputBindingsRaw] = React.useState(() => inputBindingsModule.createBindings())
+  const [root] = React.useState<React.RefObject<HTMLElement>>(() => ({
```
super cursed but i guess it makes sense...

MrFlashAccount replied:
Just an optimization to not query the element on each render.

somebody1234 commented on app/ide-desktop/lib/dashboard/src/components/AriaComponents/Button/Button.tsx:
```diff
@@ -0,0 +1,73 @@
+  readonly variant: 'icon' | 'primary'
+  readonly icon?: string
+  /**
+   * FIXME: This is not yet implemented
```
shouldn't have `FIXME`s in final code. either remove functionality or open an issue and reformat to:
```
// FIXME [author initials]: https://github.com/link/to/issue
```

somebody1234 commented on app/ide-desktop/lib/dashboard/src/components/AriaComponents/Button/Button.tsx:
```diff
@@ -0,0 +1,73 @@
+/**
+ * @file Button.tsx
```
need descriptive `@file` description

MrFlashAccount replied:
This file is descriptive enough, Especially comparing to the original Button component

somebody1234 commented on app/ide-desktop/lib/dashboard/src/components/AriaComponents/Button/Button.tsx:
```diff
@@ -0,0 +1,73 @@
+const ICON_CLASSES = 'opacity-50 hover:opacity-100'
+const EXTRA_CLICK_ZONE_CLASSES =
+  'flex relative before:inset-[-12px] before:absolute before:z-10 before:content-""'
+// not yet implemented
```
should probably be a FIXME

somebody1234 commented on app/ide-desktop/lib/dashboard/src/components/AriaComponents/Button/Button.tsx:
```diff
@@ -0,0 +1,73 @@
+  'focus-visible:outline-offset-2 focus:outline-none focus-visible:outline focus-visible:outline-blue-500 '
+const ICON_CLASSES = 'opacity-50 hover:opacity-100'
+const EXTRA_CLICK_ZONE_CLASSES =
+  'flex relative before:inset-[-12px] before:absolute before:z-10 before:content-""'
```
`before:content-""` is unnecessary. (does it even work?)

MrFlashAccount replied:
Agree, thanks for noticing!

somebody1234 commented on app/ide-desktop/lib/dashboard/src/components/AriaComponents/Button/Button.tsx:
```diff
@@ -0,0 +1,73 @@
+const DEFAULT_CLASSES =
+  'flex cursor-pointer rounded-sm border border-transparent transition-opacity duration-200 ease-in-out'
+const FOCUS_CLASSES =
+  'focus-visible:outline-offset-2 focus:outline-none focus-visible:outline focus-visible:outline-blue-500 '
```
not sure about `focus-visible:outline-blue-500` instead of `focus-visible:outline-primary`

somebody1234 commented on app/ide-desktop/lib/dashboard/src/components/AriaComponents/Button/Button.tsx:
```diff
@@ -0,0 +1,73 @@
+  const { variant, className, children, icon, ...ariaButtonProps } = props
+
+  const classes = clsx(DEFAULT_CLASSES, DISABLED_CLASSES, FOCUS_CLASSES, {
+    [ICON_CLASSES]: variant === 'icon',
```
consider using `variant === 'icon' && ICON_CLASSES` instead

MrFlashAccount replied:
Sorry, could you elaborate? I didn't get why I should use `ariant === 'icon' && ICON_CLASSES`

somebody1234 replied:
well, is there a reason to *not* use it? imo:
- we should at least be consistent in the call. either everything should be in the object or nothing should be
- `[ICON_CLASSES]: variant === 'icon'` is not as intuitive (doesn't read as naturally) as `variant === 'icon' && ICON_CLASSES` imo.

MrFlashAccount replied:
If you don't like it, it doesn't mean it's bad. And it's absolutely _natural_: https://github.com/lukeed/clsx?tab=readme-ov-file#usage

somebody1234 replied:
it's not *natural*, it's *a usecase they support*. because they support mixing and matching of every input format. that doesn't necessarily mean we want to randomly mix and match input formats, however.

MrFlashAccount replied:
The same way it doesn't mean that it's bad though.

somebody1234 commented on app/ide-desktop/lib/dashboard/src/components/AriaComponents/Button/Button.tsx:
```diff
@@ -0,0 +1,73 @@
+  })
+
+  const childrenFactory = () => {
+    return typeof icon !== 'undefined' ? (
```
`icon != null` should work fine here

somebody1234 commented on app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/index.ts:
```diff
@@ -0,0 +1 @@
+export * from './Dialog'
```
why is this component being split across multiple files?

MrFlashAccount replied:
Because large files are harder to maintain

somebody1234 replied:
while that's fair, i'm not sure a *21 line* file and a *1 line* file, *plus* an extra directory, are easier to maintain.

MrFlashAccount replied:
We've encountered a challenge with numerous files exceeding 500 lines, making maintenance more cumbersome. Originally, these files were smaller but expanded over time. Drawing from my experience implementing Dialog multiple times, I anticipate similar growth for this component. Therefore, I proactively divided a single file into multiple ones.

somebody1234 replied:
sure, but the issue is that (afaict) there is no benefit from extracting out the types and the barrel export:
- the barrel export will never be more than (say) 20 lines. since it will never contain logic, only re-exports
- the types should only ever contain the types for the prop. types for utility logic belong inside the component and should not be exported from any module, otherwise they would clutter auto-imports

somebody1234 replied:
> numerous files exceeding 500 lines

here is a comprehensive list of files > 500 lines in `src/`:
```
682 "./authentication/cognito.ts"
792 "./components/dashboard/AssetRow.tsx"
812 "./layouts/Chat.tsx"
2521 "./layouts/AssetsTable.tsx"
533 "./pages/dashboard/Dashboard.tsx"
766 "./providers/AuthProvider.tsx"
974 "./services/RemoteBackend.ts"
1210 "./services/Backend.ts"
507 "./utilities/AssetQuery.ts"
740 "./utilities/inputBindings.ts"
518 "./tailwind.css"
```
of these:
- `tailwind.css` is a css file, not sure it makes sense to split that one.
- `RemoteBackend` and `Backend` are large because they are classes that contain many (very small!) methods. again not sure it makes sense to split these.
- `AuthProvider` is similar, the only difference being that it is not a class.
- your point is valid, however, for `Dashboard`, `AssetsTable`, `AssetRow` and `Chat` - but do note:
  - i'm not sure these count as "numerous" - these *are* the most complex components *in the entire codebase* - most other components are safely well below 500 lines.
  - again - i'm not sure a `Dialog` will ever have complexity like this.

MrFlashAccount replied:
> i'm not sure these count as "numerous" - these are the most complex components in the entire codebase - most other components are safely well below 500 lines.

These components violates SRP principle, thus saying "these are the most complex components" is incorrect - they aren't complex, they just badly structured.

MrFlashAccount replied:
I'm not going to put all the Dialog logic into a single file.

somebody1234 commented on app/ide-desktop/lib/dashboard/src/components/Portal/PortalProvider.ts:
```diff
@@ -0,0 +1,42 @@
+  )
+
+  // this is safe because we are using `invariant` to check if the `PortalProvider` is in the component tree
+  // and the root element is not null
```
capitalization, `null` -> `` `null` ``

somebody1234 commented on app/ide-desktop/lib/dashboard/src/components/Portal/PortalProvider.ts:
```diff
@@ -0,0 +1,42 @@
+export function useStrictPortalContext() {
+  const root = React.useContext(PortalContext)
+
+  invariant(
```
why not just `throw new Error`?

MrFlashAccount replied:
invariant is a nice wrapper around `throw new Error`.

somebody1234 replied:
while that's fair, i'm not sure it's worth pulling in a dependency + making the code less obvious unless you already know what `invariant` does.

most importantly: it's not like it's even any faster to read, nor is it much shorter

MrFlashAccount replied:
invariant is a standard thing in js ecosystem: https://www.npmjs.com/package/tiny-invariant. Also it matches typescript's `assert` pattern: https://blog.logrocket.com/assertion-functions-typescript/

somebody1234 replied:
it's *popular*, yes. but that doesn't mean it's *standard*, nor does it mean we have to use it just because it exists.
see for example other libraries that we don't use:
<https://www.npmjs.com/package/lodash>
<https://www.npmjs.com/package/jquery>
<https://www.npmjs.com/package/@sinclair/typebox>

in general: i'm not sure about pulling in utility functions that don't really help make the code clearer

somebody1234 commented on app/ide-desktop/lib/dashboard/src/layouts/AssetPanel.tsx:
```diff
@@ -139,7 +139,7 @@ export default function AssetPanel(props: AssetPanelProps) {
               dispatchAssetEvent={dispatchAssetEvent}
             />
           )}
-          <AssetVersions hidden={tab !== AssetPanelTab.versions} item={item} />
```
this was intentional, without it being always visible, then the versions take a while to load.

MrFlashAccount replied:
And yet it doesn't take a while to load, also we store the current tab in localstorage, so it doesn't prevent users from seeing a loader. If we decide that we want to see versions _earlier_ , we might want to use prefetching(https://tanstack.com/query/latest/docs/framework/react/guides/prefetching#prefetch-in-components)

somebody1234 commented on app/ide-desktop/lib/dashboard/src/services/remoteBackendPaths.ts:
```diff
@@ -61,8 +61,16 @@ export const CREATE_CHECKOUT_SESSION_PATH = 'payments/checkout-sessions'
 /** Relative HTTP path to the "list asset versions" endpoint of the Cloud backend API. */
 export function listAssetVersionsPath(assetId: backend.AssetId) {
-  return `assets/${assetId}/versions`
+  return `assets/${assetId}/versions` as const
```
:scream: why is the `as const` needed here?

somebody1234 reviewed (approved):
review mostly finished but not completely finished

somebody1234 commented:
also not sure about the PascalCase name for `AriaComponents`

MrFlashAccount committed:
fix review issues

MrFlashAccount committed:
fix invariant

MrFlashAccount committed:
fix eslint

indiv0 commented:
Will tackle QA on this PR.

indiv0 commented:
- PR description missing screenshots and/or screencast showcasing changed behaviour.

indiv0 commented:
:x: QA failed

- ✅ Versions asset panel view now has a button to show diffs
- ✅ Diff button is disabled for the `HEAD` version
    - Maybe we shouldn't show it at all? On the one hand consistency, on the other hand took me a second to figure out why it's disabled 🤷 
    - Or alternatively, just display the contents of the `HEAD` version, without a diff?
- ✅ Clicking diff button shows a diff of the previous and the current version
- ✅ Diff page highlights changed lines in the source file
- :x: Possible to select text within the diff view
    - :x: See **Select diff text bug**
- Suggestion: the view should be themed similarly to the rest of the application. The pure white is a bit jarring. Do we have a figma anywhere for its design?
- Doesn't look like there's syntax highlighting. Is there a plan to add it?
- Unrelated to this PR, but it's super annoying that double-clicking another type of asset (e.g., a connector) opens up the asset panel, but double clicking a project does not. @somebody1234 is there an issue to track fixing that, or a reason why that is?

Side-note: this PR is awesome. Actually being able to see changes between versions, especially without having to open the project? Hell yeah 🤩 

<details>
<summary><b>Select diff text bug</b></summary>

Attempting to select text within the view breaks if selecting text that is above the asset table header.

Steps to reproduce:

1. Open Versions panel for project with multiple versions.

2. Press the diff button to open the editor.

3. Move your cursor to around where the asset table header should be.

4. Try to select some text.

5. Observe that the diff view closes unexpectedly.

https://github.com/enso-org/enso/assets/641547/d2c93592-8290-437e-88bc-0db5e3351ab0

</details>

MrFlashAccount committed:
fix prettier

MrFlashAccount committed:
fix selection inside Monaco Editor

MrFlashAccount commented:
@indiv0 thank you for such a thorough qa! I fixed the bug you mentioned.

Addressing your suggestions:
>the view should be themed similarly to the rest of the application. The pure white is a bit jarring. Do we have a figma anywhere for its design?

No, we don't have design yet, not sure it's worth spending time atm though.

> Doesn't look like there's syntax highlighting. Is there a plan to add it?

Yea, we had that, but eventually moved this out of scope cause it requires some refactoring of our vscode plugin(we need to put the syntax schema somewhere else to use it in dashboard), Hope eventually we'll add it!

somebody1234 commented:
> we don't have design yet

note though, that we should at least make sure things like background color, backdrop filter, heading size, heading color and the like are consistent with all other modals

indiv0 commented:
@MrFlashAccount No problem :)

> Yea, we had that, but eventually moved this out of scope cause it requires some refactoring of our vscode plugin(we need to put the syntax schema somewhere else to use it in dashboard), Hope eventually we'll add it!

Okay sounds good. Would you mind opening an issue to track it? Either here or in the cloud-v2 repo. We'll put it in the backlog for now.

> No, we don't have design yet, not sure it's worth spending time atm though.

If we don't have a design, then no need to spend time making one. I agree with e-hern though, the theming should at least match the existing modals.

Ping me once the theme is updated and I'll re-QA :)

MrFlashAccount commented:
@somebody1234 @indiv0 

![CleanShot 2024-03-22 at 18 00 28@2x](https://github.com/enso-org/enso/assets/61194245/3a36e94b-6c9f-4414-9bd4-0fd28e9a014c)

I did a few fixes that should make this dialog to look closer to the others dialogs. I left the title larger though, because small title looks weird with full-screen dialog.

MrFlashAccount committed:
fix bg

indiv0 commented:
Minor note: it's a bit odd that the size of the view extends past the end of the viewport. That is, it's unexpected to see a scrollbar when viewing a small file.

indiv0 commented:
QA passed! ✅
